ocaml/xapi/repository.ml

```diff
@@ -512,50 +512,53 @@ let get_host_updates_in_json ~__context ~installed =
         (ExnHelper.string_of_exn e) ;
       raise Api_errors.(Server_error (get_host_updates_failed, [ref]))
 
+(* Map the URI in the HTTP reuqest to local repo path *)
+let map_uri_to_path ~prefix ~by uri =
+  let regex = prefix |> Re.str |> Re.compile in
+  uri
+  |> Uri.of_string
+  |> Uri.path
+  |> Re.replace_string ~all:false regex ~by
+  |> Helpers.resolve_uri_path ~root:!Xapi_globs.local_pool_repo_dir
+
 let get_repository_handler (req : Http.Request.t) s _ =
   let open Http in
-  let open Xapi_stdext_std.Xstringext in
-  debug "Repository.get_repository_handler URL %s" req.Request.uri ;
+  debug "%s URL: %s" __FUNCTION__ req.Request.uri ;
   req.Request.close <- true ;
   if Fileserver.access_forbidden req s then
     Http_svr.response_forbidden ~req s
   else
-    let can_be_authorized =
-      try
-        Xapi_http.with_context "get_repository_handler" req s (fun _ -> ()) ;
-        true
-      with _ -> false
-    in
-    let internal_repo_access_only =
-      let __context =
-        Context.make ~origin:(Http (req, s)) "get_repository_handler"
-      in
-      Pool_features.is_enabled ~__context Features.Internal_repo_access
-    in
-    match (can_be_authorized, internal_repo_access_only) with
-    | false, true ->
-        error
-          "Invalid secret for authorization when Internal_repo_access is \
-           enabled" ;
-        Http_svr.response_forbidden ~req s
-    | _ -> (
-      try
-        let len = String.length Constants.get_repository_uri in
-        match String.sub_to_end req.Request.uri len with
-        | uri_path ->
-            let root = !Xapi_globs.local_pool_repo_dir in
-            Fileserver.response_file s (Helpers.resolve_uri_path ~root ~uri_path)
-        | exception e ->
-            let msg =
-              Printf.sprintf "Failed to get path from uri': %s"
-                (ExnHelper.string_of_exn e)
-            in
-            raise Api_errors.(Server_error (internal_error, [msg]))
-      with e ->
-        error "Failed to serve for request on uri %s: %s" req.Request.uri
-          (ExnHelper.string_of_exn e) ;
-        Http_svr.response_forbidden ~req s
-    )
+    try
+      req.Request.uri
+      |> map_uri_to_path ~prefix:Constants.get_repository_uri ~by:String.empty
+      |> Fileserver.response_file s
+    with e ->
+      error "Failed to serve request %s with exception: %s" req.Request.uri
+        (ExnHelper.string_of_exn e) ;
+      Http_svr.response_forbidden ~req s
+
+let get_enabled_repository_handler (req : Http.Request.t) s _ =
```

Comment:
Can't the two handlers `get_repository_handler` and `get_enabled_repository_handler` share the same bits?